### PR TITLE
Pro 6829 extend caveat from bulk scan

### PR DESCRIPTION
**Pull request by jackmunkonge**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PRO-6829

### Change description ###
Added a new endpoint ('/update-case') to be used by the Bulk Scanning team for extending a caveat if a PA8A form has been completed with the 'caseReference' field populated. The request from bulk scanning should have the CCD case included, so this branch does the following checks.
1. Verifies this form-type is a PA8A 
2. Verifiies caseReference is present in the OCR and matches the attached case details.
3. Moves scanned documents from the exception record into the case details, but ignores DCN's which already exist.
4. Calls the existing validation routines for the Caveat extension to verify the correct data in the case.
5. Updates the Caveat extension expiry date by adding 6 months onto the existing date.
6. Calls the routines for notifying the applicant once the caveat extension has been changed.

All of this is done before returning this caveat details back to the Bulk Scanning process to be committed to CCD. 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by davejones74 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/probate/service/exceptionrecord/ExceptionRecordService.java [hunk 1]
```diff
@@ -114,4 +140,104 @@ public SuccessfulTransformationResponse createGrantOfRepresentationCaseFromExcep
...
+                    .build();
+
+        } catch(Exception e){
+            log.error("Error Extending Caveat case from Exception Record", e);
+            throw new OCRMappingException(e.getMessage());
+        }
+    }
+
+    private List<uk.gov.hmcts.probate.model.ccd.raw.CollectionMember<uk.gov.hmcts.probate.model.ccd.raw.ScannedDocument>> mergeScannedDocuments(
+            List<uk.gov.hmcts.probate.model.ccd.raw.CollectionMember<uk.gov.hmcts.probate.model.ccd.raw.ScannedDocument>> caseScannedDocuments,
```
Comment (davejones74) [inline src/main/java/uk/gov/hmcts/probate/service/exceptionrecord/ExceptionRecordService.java:219]: caseScannedDocuments.size() needs to be compared with newScannedDocuments.size() after the merge. If both are the same then Bulk Scanning have told us they will not update. So I think an OCRValidationExcption needs to be thrown if this is the case with an appropriate message to halt the update.

File: src/main/java/uk/gov/hmcts/probate/service/exceptionrecord/ExceptionRecordService.java [hunk 2]
```diff
@@ -114,4 +140,104 @@ public SuccessfulTransformationResponse createGrantOfRepresentationCaseFromExcep
             throw new OCRMappingException(e.getMessage());
         }
     }
+
+    public SuccessfulCaveatUpdateResponse updateCaveatCaseFromExceptionRecord(
```
Comment (davejones74) [inline src/main/java/uk/gov/hmcts/probate/service/exceptionrecord/ExceptionRecordService.java:145]: The class updateCaveatCaseFromExceptionRecord needs a unit test.

File: src/main/java/uk/gov/hmcts/probate/controller/ExceptionRecordController.java [hunk 3]
```diff
@@ -120,9 +122,56 @@
...
 
-    private void logCallback(SuccessfulTransformationResponse callbackResponse) {
+    @ApiOperation(value = "Updates a case based on availability of OCR data and documents", notes = "Will return errors if unsuccessful or no new documents found.")
+    @ApiResponses({
+            @ApiResponse(code = 200, response = ValidationResponse.class, message = "Validation executed successfully"),
+            @ApiResponse(code = 400, message = "Request failed due to malformed syntax"),
+            @ApiResponse(code = 403, message = "S2S token is not authorized, missing or invalid"),
+            @ApiResponse(code = 404, message = "Form type not found")
+    })
+    @PostMapping(path = "/update-case",
```
Comment (davejones74) [inline src/main/java/uk/gov/hmcts/probate/controller/ExceptionRecordController.java:132]: This controller method needs a unit test.

File: src/main/java/uk/gov/hmcts/probate/transformer/CaveatCallbackResponseTransformer.java [hunk 4]
```diff
@@ -153,7 +155,8 @@ public CaveatCallbackResponse addMatches(CaveatCallbackRequest request, List<Cas
     public CaveatCallbackResponse transformResponseWithExtendedExpiry(CaveatCallbackRequest caveatCallbackRequest) {
```
Comment (davejones74) [inline src/main/java/uk/gov/hmcts/probate/transformer/CaveatCallbackResponseTransformer.java:154]: The method transformResponseWithExtendedExpiry needs a unit test.

**Review by davejones74 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/probate/service/exceptionrecord/ExceptionRecordService.java (hunk 1 above)
Comment (sanjayparekh-hmcts) [reply src/main/java/uk/gov/hmcts/probate/service/exceptionrecord/ExceptionRecordService.java:219]: done on my commit - needs checking

File: src/main/java/uk/gov/hmcts/probate/service/exceptionrecord/ExceptionRecordService.java (hunk 2 above)
Comment (sanjayparekh-hmcts) [reply src/main/java/uk/gov/hmcts/probate/service/exceptionrecord/ExceptionRecordService.java:145]: done on my commit - needs checking

File: src/main/java/uk/gov/hmcts/probate/controller/ExceptionRecordController.java (hunk 3 above)
Comment (sanjayparekh-hmcts) [reply src/main/java/uk/gov/hmcts/probate/controller/ExceptionRecordController.java:132]: added tests

File: src/main/java/uk/gov/hmcts/probate/transformer/CaveatCallbackResponseTransformer.java (hunk 4 above)
Comment (sanjayparekh-hmcts) [reply src/main/java/uk/gov/hmcts/probate/transformer/CaveatCallbackResponseTransformer.java:154]: added

**Review by davejones74 — APPROVED** [review]

**Pull request merged**